File: 11_Bombardment/javascript/bombardment.js
// BOMBARDMENT
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(33) + "BOMBARDMENT\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("YOU ARE ON A BATTLEFIELD WITH 4 PLATOONS AND YOU\n");
    print("HAVE 25 OUTPOSTS AVAILABLE WHERE THEY MAY BE PLACED.\n");
    print("YOU CAN ONLY PLACE ONE PLATOON AT ANY ONE OUTPOST.\n");
    print("THE COMPUTER DOES THE SAME WITH ITS FOUR PLATOONS.\n");
    print("\n");
    print("THE OBJECT OF THE GAME IS TO FIRE MISSILES AT THE\n");
    print("OUTPOSTS OF THE COMPUTER.  IT WILL DO THE SAME TO YOU.\n");
    print("THE ONE WHO DESTROYS ALL FOUR OF THE ENEMY'S PLATOONS\n");
    print("FIRST IS THE WINNER.\n");
    print("\n");
    print("GOOD LUCK... AND TELL US WHERE YOU WANT THE BODIES SENT!\n");
    print("\n");
    // "TEAR OFF" because it supposed this to be printed on a teletype
    print("TEAR OFF MATRIX AND USE IT TO CHECK OFF THE NUMBERS.\n");
    for (r = 1; r <= 5; r++)
        print("\n");
    ma = [];
    for (r = 1; r <= 100; r++)
        ma[r] = 0;
    p = 0;
    q = 0;
    z = 0;
    for (r = 1; r <= 5; r++) {
        i = (r - 1) * 5 + 1;
        print(i + "\t" + (i + 1) + "\t" + (i + 2) + "\t" + (i + 3) + "\t" + (i + 4) + "\n");
    }
    for (r = 1; r <= 10; r++)
        print("\n");
    c = Math.floor(Math.random() * 25) + 1;
    do {
        d = Math.floor(Math.random() * 25) + 1;
        e = Math.floor(Math.random() * 25) + 1;
        f = Math.floor(Math.random() * 25) + 1;
    } while (c == d || c == e || c == f || d == e || d == f || e == f) ;
    print("WHAT ARE YOUR FOUR POSITIONS");
    str = await input();
    g = parseInt(str);
    str = str.substr(str.indexOf(",") + 1);
    h = parseInt(str);
    str = str.substr(str.indexOf(",") + 1);
    k = parseInt(str);
    str = str.substr(str.indexOf(",") + 1);
    l = parseInt(str);
    print("\n");
    // Another "bug" your outpost can be in the same position as a computer outpost
    // Let us suppose both live in a different matrix.
    while (1) {
        // The original game didn't limited the input to 1-25
        do {
            print("WHERE DO YOU WISH TO FIRE YOUR MISSLE");
            y = parseInt(await input());
        } while (y < 0 || y > 25) ;
        if (y == c || y == d || y == e || y == f) {
            
            // The original game has a bug. You can shoot the same outpost
            // several times. This solves it.
            if (y == c)
                c = 0;
            if (y == d)
                d = 0;
            if (y == e)
                e = 0;
            if (y == f)
                f = 0;
            q++;
            if (q == 1) {
                print("ONE DOWN. THREE TO GO.\n");
            } else if (q == 2) {
                print("TWO DOWN. TWO TO GO.\n");
            } else if (q == 3) {
                print("THREE DOWN. ONE TO GO.\n");
            } else {
                print("YOU GOT ME, I'M GOING FAST. BUT I'LL GET YOU WHEN\n");
                print("MY TRANSISTO&S RECUP%RA*E!\n");
                break;
            }
        } else {
            print("HA, HA YOU MISSED. MY TURN NOW:\n");
        }
        print("\n");
        print("\n");
        do {
            m = Math.floor(Math.random() * 25 + 1);
            p++;
            n = p - 1;
            for (t = 1; t <= n; t++) {
                if (m == ma[t])
                    break;
            }
        } while (t <= n) ;
        x = m;
        ma[p] = m;
        if (x == g || x == h || x == l || x == k) {
            z++;
            if (z < 4)
                print("I GOT YOU. IT WON'T BE LONG NOW. POST " + x + " WAS HIT.\n");
            if (z == 1) {
                print("YOU HAVE ONLY THREE OUTPOSTS LEFT.\n");
            } else if (z == 2) {
                print("YOU HAVE ONLY TWO OUTPOSTS LEFT.\n");
            } else if (z == 3) {
                print("YOU HAVE ONLY ONE OUTPOST LEFT.\n");
            } else {
                print("YOU'RE DEAD. YOUR LAST OUTPOST WAS AT " + x + ". HA, HA, HA.\n");
                print("BETTER LUCK NEXT TIME.\n");
            }
        } else {
            print("I MISSED YOU, YOU DIRTY RAT. I PICKED " + m + ". YOUR TURN:\n");
        }
        print("\n");
        print("\n");
    }
}

main();

File: 11_Bombardment/ruby/bombardment.rb
require "set"

class Battlefield
  TOTAL_OUTPOSTS = 25
  TOTAL_PLATOONS = 4

  class << self
    def with_random_platoon_placements
      new((1..TOTAL_OUTPOSTS).to_a.sample(4))
    end

    def valid_platoon_placement?(i)
      i.between?(1, TOTAL_OUTPOSTS)
    end
  end

  def initialize(platoon_placements)
    @platoon_placements = Set.new(platoon_placements)
  end

  # Fires a missile at the specified outpost. Returns true if it hits a
  # platoon, false otherwise.
  def fire_missile_at(outpost)
    !@platoon_placements.delete?(outpost).nil?
  end

  def remaining_platoons_count
    @platoon_placements.size
  end
end

class Game
  def initialize(player_battlefield:, computer_battlefield:)
    @player_battlefield = player_battlefield
    @computer_battlefield = computer_battlefield
  end

  def play
    player_choice = get_integer_answer_to("WHERE DO YOU WISH TO FIRE")

    if @computer_battlefield.fire_missile_at(player_choice)
      print_player_success_message
      return if @computer_battlefield.remaining_platoons_count == 0
    else
      puts "HA, HA YOU MISSED. MY TURN NOW:"
    end
    puts

    computer_choice = rand(1..Battlefield::TOTAL_OUTPOSTS)

    if @player_battlefield.fire_missile_at(computer_choice)
      print_computer_success_message(computer_choice)
    else
      puts "I MISSED YOU, YOU DIRTY RAT. I PICKED #{computer_choice}. YOUR TURN:"
    end
    puts
  end

  def finished?
    @player_battlefield.remaining_platoons_count == 0 ||
      @computer_battlefield.remaining_platoons_count == 0
  end

  private

  def get_integer_answer_to(question)
    puts question
    $stdin.gets.strip.to_i
  end

  def print_computer_success_message(computer_choice)
    if @player_battlefield.remaining_platoons_count == 0
      puts "YOU'RE DEAD. YOUR LAST OUTPOST WAS AT #{computer_choice}. HA, HA, HA."
      puts "BETTER LUCK NEXT TIME."
      return
    end

    puts "I GOT YOU. IT WON'T BE LONG NOW. POST #{computer_choice} WAS HIT."

    case @player_battlefield.remaining_platoons_count
    when 1
      puts "YOU HAVE ONLY ONE OUTPOST LEFT"
    when 2
      puts "YOU HAVE ONLY TWO OUTPOSTS LEFT"
    when 3
      puts "YOU HAVE ONLY THREE OUTPOSTS LEFT"
    end
  end

  def print_player_success_message
    if @computer_battlefield.remaining_platoons_count == 0
      puts "YOU GOT ME, I'M GOING FAST. BUT I'LL GET YOU WHEN"
      puts "MY TRANSISTO&S RECUP%RA*E!"
      return
    end

    puts "YOU GOT ONE OF MY OUTPOSTS!"

    case @computer_battlefield.remaining_platoons_count
    when 1
      puts "THREE DOWN, ONE TO GO."
    when 2
      puts "TWO DOWN, TWO TO GO."
    when 3
      puts "ONE DOWN, THREE TO GO."
    end
  end
end

INTRODUCTION = <<~TEXT
                                   BOMBARDMENT
                 CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY



  YOU ARE ON A BATTLEFIELD WITH 4 PLATOONS AND YOU
  HAVE 25 OUTPOSTS AVAILABLE WHERE THEY MAY BE PLACED.
  YOU CAN ONLY PLACE ONE PLATOON AT ANY ONE OUTPOST.
  THE COMPUTER DOES THE SAME WITH ITS FOUR PLATOONS.

  THE OBJECT OF THE GAME IS TO FIRE MISSLES AT THE
  OUTPOSTS OF THE COMPUTER.  IT WILL DO THE SAME TO YOU.
  THE ONE WHO DESTROYS ALL FOUR OF THE ENEMY'S PLATOONS
  FIRST IS THE WINNER.

  GOOD LUCK... AND TELL US WHERE YOU WANT THE BODIES SENT!

  TEAR OFF MATRIX AND USE IT TO CHECK OFF THE NUMBERS.
TEXT

def print_matrix
  (1..25).each do |i|
    print i
    print i % 5 == 0 ? "\n" : "\t"
  end
end

def ask_for_platoon_placements
  puts "WAHT ARE YOUR FOUR POSITIONS?"
  platoon_placements = $stdin.gets.strip.split(",").map(&:to_i)

  if platoon_placements.size == 4 &&
      platoon_placements.all? { |i| Battlefield.valid_platoon_placement?(i) }
    platoon_placements
  else
    puts "POSITIONS SHOULD BE BETWEEN 1 TO 25, DELIMITED BY COMMA."
    ask_for_platoon_placements
  end
end

def start_game
  puts INTRODUCTION
  puts
  print_matrix
  puts
  platoon_placements = ask_for_platoon_placements
  puts

  game = Game.new(
    player_battlefield: Battlefield.new(platoon_placements),
    computer_battlefield: Battlefield.with_random_platoon_placements
  )

  game.play until game.finished?
end

start_game
